lib/report-translator.js

```diff
@@ -216,6 +211,12 @@ function createProblem(options) {
         source: options.sourceLines[options.loc.start.line - 1] || ""
     };
 
+    // If this isn’t in the conditional, some of the tests fail
+    // because `messageId` is present in the problem object
+    if (options.messageId) {
+        problem.messageId = options.messageId;
```

Comment:
I'm confused -- do we want to include the `messageId` in the problem, or not? (Also cc'ing @platinumazure, since this is more of a design question)

Sorry if it seems like I'm going back and forth on this. This change adds some complexity to the rule API, and it will be difficult to make modifications to the API later without breaking compatibility. So I would feel better about the change knowing that (a) it provides benefits that clearly outweigh the additional complexity costs, and (b) we've thoroughly discussed the design details to ensure that those benefits will actually be accessible with the API that we expose.

I'm assuming you added `problem.messageId` in response to my suggestion about integrations in https://github.com/eslint/eslint/pull/9165#discussion_r135396568. Thanks for doing that, although I'm unsure whether it's useful for an integration to have access to the `messageId` if they don't also have access to the data that was used as part of message interpolation, since that data would be needed to reconstruct the message or use i18n.

I'm a bit unclear about what the design goal of this feature is (maybe @platinumazure can help explain). From what I can tell (feel free to correct me if I describe anything incorrectly), one of the following is the case:

1. We want i18n to be handled by plugins and rule authors. In that case, it seems like message IDs are an internal detail and don't need to be exposed as part of the final problem object. (I'm unsure whether `messageIds` provide a significant a benefit over manual string concatenation in this case, although there might be something I'm missing.)
1. We want i18n to be handled by integrations. In that case, we would probably need to let the integrations handle message interpolation as well, otherwise they won't have enough info to translate messages.

I'm not trying to stand in the way here -- I just want to make sure we're communicating the design goals clearly, so that the resulting API is coherent and useful.

Reply:
How does this sound?

1. Remove the `messageId` from the report
2. Open an issue to allow overriding the message-generating infrastructure
3. Allow the values in the `messages` object to be functions that get called to retrieve the value. Usage:
    ```js
    module.exports = {
      meta: {
        messages: {
          foo: (optionalDataParam) => myI18nLib('message here')
        }
      }
    }
    ```

Reply:
That seems like it could work, although I'd also like to hear what @platinumazure thinks.

Reply:
@platinumazure Any comments on this design?

Reply:
@platinumazure Friendly ping

Reply:
Ack, I didn't see this thread when I commented earlier today. Something like the proposed design could also work, but I'm a fan of some sort of core-level hook that would allow i18n/l10n to be separate from rules, if desired.